### ISSUE TITLE
Unable to load image asset on low-res devices if only high-resolution variants exist

### ISSUE THREAD

**Issue by yeshu-cn** [flutter/flutter#127090]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

## Description:
After upgrading to Flutter 3.10.1, I've noticed that low resolution devices are unable to correctly utilize high resolution image assets. Within my image resources directory, I have a 3.0x folder that contains the icon_apps.png image.

In Flutter 3.7.12, when I utilized image_resolution on a 2x device, candidateVariants would correctly return assets/images/3.0x/icon_apps.png. However, in Flutter 3.10.1, I observed that a 2x device would return two candidateVariants: assets/images/icon_apps.png and assets/images/3.0x/icon_apps.png.

Below is the related source code from Flutter 3.10.1 SDK for image selection:

```dart
AssetManifest.loadFromAssetBundle(chosenBundle)
      .then((AssetManifest manifest) {
        final Iterable<AssetMetadata>? candidateVariants = manifest.getAssetVariants(keyName);
        final AssetMetadata chosenVariant = _chooseVariant(
          keyName,
          configuration,
          candidateVariants,
        );
      });
```

### Steps to reproduce

- Create a project in Flutter 3.10.1, and within the assets/images/ directory, create a 3.0x/ directory, and place the icon_apps.png image in this directory.
- Declare the image assets in the pubspec.yaml file.
```yaml
flutter:
  assets:
    - assets/images/icon_apps.png
```
- On a 2x device, load the icon_apps.png image :`Image.asset("asset/images/icon_apps.png"),`.

### Expected results

On low resolution devices, even if the device resolution is lower than the image asset resolution, it should still be able to correctly utilize high resolution image assets.

### Actual results

On low resolution devices, it's not possible to utilize high resolution image assets because image_resolution returns incorrect candidateVariants and chooses the wrong variant.

### Code sample

asset dir files:
```
assets
    - images
        - 3.0x
            - icon_apps.png 
```
pubsepc.yaml
```yaml
flutter:
  assets:
    - assets/images/icon_apps.png
```

```dart
Image.asset("asset/images/icon_apps.png"),
```

</details>

### Logs

Exception: Asset not found
#0      PlatformAssetBundle.loadBuffer (package:flutter/src/services/asset_bundle.dart:369:7)
<asynchronous suspension>
#1      AssetBundleImageProvider._loadAsync (package:flutter/src/painting/image_provider.dart:811:18)
<asynchronous suspension>

### Flutter Doctor output

$ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.10.1, on macOS 13.0.1 22A400 darwin-arm64, locale zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.2)
[✓] IntelliJ IDEA Community Edition (version 2022.2.3)
[✓] VS Code (version 1.78.1)
[✓] Connected device (3 available)
[✓] Network resources

• No issues found!
(base)

**Comment by huycozy** [comment]
Hi @yeshu-cn 
Can you share which devices are you testing on this? And please provide the complete output of `flutter run -v` as well. Thank you!

**Comment by yeshu-cn** [comment]
Device: iPhone SE,14.0

Fluttter run -v：
```
[ +182 ms] (lldb) command source -s 0 '/tmp/E3121990-A96A-4C2E-AA21-5F5DBC8DCC39/fruitstrap-lldb-prep-cmds-00008030_001A2C803C43802E'
[        ] Executing commands in '/tmp/E3121990-A96A-4C2E-AA21-5F5DBC8DCC39/fruitstrap-lldb-prep-cmds-00008030_001A2C803C43802E'.
[        ] (lldb)     platform select remote-'ios' --sysroot '/Users/yeshu/Library/Developer/Xcode/iOS DeviceSupport/14.0 (18A373) arm64e/Symbols'
[        ]   Platform: remote-ios
[        ]  Connected: no
[        ]    Sysroot: /Users/yeshu/Library/Developer/Xcode/iOS DeviceSupport/14.0 (18A373) arm64e/Symbols
[        ]   SDK Path: "/Users/yeshu/Library/Developer/Xcode/iOS DeviceSupport/14.0 (18A373) arm64e/Symbols"
[        ] (lldb)     target create "/Users/yeshu/Downloads/flutter/flutterdemo/build/ios/iphoneos/Runner.app"
[+4772 ms] Current executable set to '/Users/yeshu/Downloads/flutter/flutterdemo/build/ios/iphoneos/Runner.app' (arm64).
[        ] (lldb)     script fruitstrap_device_app="/private/var/containers/Bundle/Application/36B2052D-C4D4-4FB7-B86D-73DC054F5C80/Runner.app"
[ +106 ms] (lldb)     script fruitstrap_connect_url="connect://127.0.0.1:53755"
[        ] (lldb)     script fruitstrap_output_path=""
[        ] (lldb)     script fruitstrap_error_path=""
[        ] (lldb)     target modules search-paths add /usr "/Users/yeshu/Library/Developer/Xcode/iOS DeviceSupport/14.0 (18A373) arm64e/Symbols/usr" /System "/Users/yeshu/Library/Developer/Xcode/iOS DeviceSupport/14.0 (18A373) arm64e/Symbols/System" "/private/var/containers/Bundle/Application/36B2052D-C4D4-4FB7-B86D-73DC054F5C80" "/Users/yeshu/Downloads/flutter/flutterdemo/build/ios/iphoneos" "/var/containers/Bundle/Application/36B2052D-C4D4-4FB7-B86D-73DC054F5C80" "/Users/yeshu/Downloads/flutter/flutterdemo/build/ios/iphoneos" /Developer "/Users/yeshu/Library/Developer/Xcode/iOS DeviceSupport/14.0 (18A373) arm64e/Symbols/Developer"
[  +38 ms] (lldb)     command script import "/tmp/E3121990-A96A-4C2E-AA21-5F5DBC8DCC39/fruitstrap_00008030_001A2C803C43802E.py"
[   +3 ms] (lldb)     command script add -f fruitstrap_00008030_001A2C803C43802E.connect_command connect
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_00008030_001A2C803C43802E.run_command run
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_00008030_001A2C803C43802E.autoexit_command autoexit
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_00008030_001A2C803C43802E.safequit_command safequit
[        ] (lldb)     connect
[  +12 ms] (lldb)     run
[ +125 ms] success
[        ] Application launched on the device. Waiting for Dart VM Service url.
[ +793 ms] (lldb) 2023-05-19 10:30:49.101205+0800 Runner[1065:467607] [VERBOSE-2:FlutterDarwinContextMetalImpeller.mm(35)] Using the Impeller rendering backend.
[  +85 ms] VM Service URL on device: http://127.0.0.1:63088/xo7oi01Jy34=/
[   +3 ms] Attempting to forward device port 63088 to host port 53772
[        ] executing: /Users/yeshu/development/flutter/bin/cache/artifacts/usbmuxd/iproxy 53772:63088 --udid 00008030-001A2C803C43802E --debug
[+1009 ms] Forwarded port ForwardedPort HOST:53772 to DEVICE:63088
[   +1 ms] Forwarded host port 53772 to device port 63088 for VM Service
[   +4 ms] Installing and launching... (completed in 9.0s)
[   +2 ms] Caching compiled dill
[  +69 ms] Connecting to service protocol: http://127.0.0.1:53772/xo7oi01Jy34=/
[ +126 ms] Launching a Dart Developer Service (DDS) instance at http://127.0.0.1:0, connecting to VM service at http://127.0.0.1:53772/xo7oi01Jy34=/.
[  +49 ms] DDS is listening at http://127.0.0.1:53777/hfpfZU0xBzM=/.
[  +25 ms] Successfully connected to service protocol: http://127.0.0.1:53772/xo7oi01Jy34=/
[  +12 ms] DevFS: Creating new filesystem on the device (null)
[  +11 ms] DevFS: Created new filesystem on the device (file:///private/var/mobile/Containers/Data/Application/EED75860-CF2E-475F-A0F9-D9CF9803A7D9/tmp/flutterdemo5DbRTB/flutterdemo/)
[        ] Updating assets
[  +75 ms] Syncing files to device iPhone的...
[   +1 ms] Compiling dart to kernel with 0 updated files
[        ] Processing bundle.
[        ] <- recompile package:flutterdemo/main.dart 89e70c57-303a-4277-8162-6df8a7cf3afd
[        ] <- 89e70c57-303a-4277-8162-6df8a7cf3afd
[   +1 ms] Bundle processing done.
[ +115 ms] Updating files.
[        ] DevFS: Sync finished
[        ] Syncing files to device iPhone的... (completed in 119ms)
[        ] Synced 0.0MB.
[        ] <- accept
[   +3 ms] Connected to _flutterView/0x10400fc20.
[   +1 ms] Flutter run key commands.
[        ] r Hot reload. 🔥🔥🔥
[        ] R Hot restart.
[        ] h List all available interactive commands.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] A Dart VM Service on iPhone的 is available at: http://127.0.0.1:53777/hfpfZU0xBzM=/
[        ] The Flutter DevTools debugger and profiler on iPhone的 is available at: http://127.0.0.1:9100?uri=http://127.0.0.1:53777/hfpfZU0xBzM=/
[+1526 ms] ══╡ EXCEPTION CAUGHT BY IMAGE RESOURCE SERVICE ╞════════════════════════════════════════════════════
           The following assertion was thrown resolving an image codec:
           Unable to load asset: "assets/images/icon_apps.png".
           Exception: Asset not found

           When the exception was thrown, this was the stack:
           #0      PlatformAssetBundle.loadBuffer (package:flutter/src/services/asset_bundle.dart:369:7)
           <asynchronous suspension>
           #1      AssetBundleImageProvider._loadAsync (package:flutter/src/painting/image_provider.dart:811:18)
           <asynchronous suspension>

           Image provider: AssetImage(bundle: null, name: "assets/images/icon_apps.png")
           Image key: AssetBundleImageKey(bundle: PlatformAssetBundle#2ffa9(), name:
             "assets/images/icon_apps.png", scale: 1.0)
           ════════════════════════════════════════════════════════════════════════════════════════════════════
[+7103 ms] Application finished.
```

This is the complete log. @huycozy 
[log.txt](https://github.com/flutter/flutter/files/11513008/log.txt)

**Comment by yeshu-cn** [comment]
I tried to look at the source code. In Flutter 3.7.12, it returned a single image path, as shown in the screenshot. 
![image](https://github.com/flutter/flutter/assets/3785459/58d98d6b-5f74-4df6-b381-4f25e88ac42e)
However, in Flutter 3.10.1, it returned two image paths, and ultimately, the `assets/images/icon_apps.png` was used.
![image](https://github.com/flutter/flutter/assets/3785459/2c8e23fe-82bb-4cdb-a23c-e407ccbd2b58)

**Comment by huycozy** [comment]
Thanks for your response. I check this issue on `iPhone SE (2nd gen)` and `iPhone 7` - 326 ppi but the image still displays as expected:

<details>
<summary>Demo - iPhone SE (2nd gen)</summary>

![Screenshot 2023-05-19 at 14 21 18](https://github.com/flutter/flutter/assets/104349824/31d1021f-d5c6-4a57-9c33-c841bd187eca)

</details>

<details>
<summary>Sample code</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(home: MyApp()));
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Image.asset('assets/images/ic_launcher.png'),
      ),
    );
  }
}
```

</details>

Assets tree:

```console
➜  assets tree -L 3
.
├── Butterfly-209.mp4
├── house.png
├── images
│   ├── 1.5x
│   │   └── ic_launcher.png
│   ├── 2.0x
│   │   └── ic_launcher.png
│   ├── 3.0x
│   │   └── ic_launcher.png
│   ├── 4.0x
│   │   └── ic_launcher.png
│   ├── ic_launcher.png
│   └── red_square.png
├── picture.png
└── red_square.png

6 directories, 10 files
```

Verifying this on the latest stable `3.10.1`. I'm not sure if I'm missing something related to sample code, can you please share the complete sample code? (maybe a zip file of project or Github repository would be appreciated)

**Comment by yeshu-cn** [comment]
```
➜  assets tree -L 3
.
├── Butterfly-209.mp4
├── house.png
├── images
│   ├── 3.0x
│   │   └── ic_launcher.png
│   └── red_square.png
├── picture.png
└── red_square.png

6 directories, 10 files
```
Delete other resolution image resource, **just keep `images/3.0x/ic_launcher.png`**。

If only high-resolution images exist, wouldn't low-resolution devices automatically use the high-resolution images? This worked fine before upgrading to Flutter 3.10.1. I'm not sure if it's a bug in Flutter 3.10.1, or if I'm using image assets incorrectly, or if the usage has changed in 3.10.1.

Thank your

**Comment by yeshu-cn** [comment]
[issues_image.zip](https://github.com/flutter/flutter/files/11514830/issues_image.zip)

run on iPhone 14 Pro Max - iOS 16.4
![image](https://github.com/flutter/flutter/assets/3785459/ee077cfe-cceb-44b2-b62f-674438534b34)


iPhone SE(3rd generation) - iOS 16.4
![image](https://github.com/flutter/flutter/assets/3785459/86bc7378-0bf1-4113-8bca-e04ad7d76013)

**Comment by huycozy** [comment]
After deleting other assets and only keeping 3.0x, I retry again on iPhone 14 Pro and on Flutter 3.7.12 but get the same issue. 
You can try my sample code here: [issue_127090_resolution_assets.zip](https://github.com/flutter/flutter/files/11515540/issue_127090_resolution_assets.zip)

I'm not sure which situation can cause only high-resolution images exist, but I think images need to be declared completely as [resolution-aware](https://docs.flutter.dev/ui/assets-and-images#resolution-aware) guideline to work.

**Comment by yeshu-cn** [comment]
> After deleting other assets and only keeping 3.0x, I retry again on iPhone 14 Pro and on Flutter 3.7.12 but get the same issue. Not sure why you get the different result, but this could be the issue of the wrong setup I think.

I have already run your sample code，you need to add `- assets/images/ic_launcher.png` to the pubspec.yaml file, then Flutter 3.7.12(iPhone SE and iphone 14) should work successfully. However, it will fail to run on 3.10.1(iphone SE), resulting in an "asset not found" error。

I have uploaded my sample code here:https://github.com/flutter/flutter/issues/127090#issuecomment-1554205303


> I'm not sure which situation can cause only high-resolution images exist, but I think images need to be declared completely as [resolution-aware](https://docs.flutter.dev/ui/assets-and-images#resolution-aware) guideline to work.

Because I only have 3x image resources, just like in Android development, UI designers generally only provide one set of images at the most commonly used resolution, such as xxhdpi.

**Comment by goderbauer** [comment]
cc @andrewkolos since you recently touched this code path.

**Comment by andrewkolos** [comment]
Thanks for bringing this to my attention.

@yeshu-cn, thank you for the astounding work you put into filing and investigating this issue. With all the information here, I was able to quickly determine that this is indeed a bug and that it was introduced by https://github.com/flutter/flutter/pull/122505. (Also cheers to @huycozy for triaging, of course 🙂).

I'll start working on a PR to fix this.

**Comment by andrewkolos** [comment]
> > I'm not sure which situation can cause only high-resolution images exist, but I think images need to be declared completely as [resolution-aware](https://docs.flutter.dev/ui/assets-and-images#resolution-aware) guideline to work.
> 
> Because I only have 3x image resources, just like in Android development, UI designers generally only provide one set of images at the most commonly used resolution, such as xxhdpi.

To clear up any remaining confusion here, it is indeed valid to declare only the "main" asset (even if it doesn't exist) as long as there exists corresponding image-resolution aware variants of that asset. From [Declaring resolution-aware image assets](https://docs.flutter.dev/ui/assets-and-images#resolution-aware):

> Each entry in the asset section of the pubspec.yaml should correspond to a real file, with the exception of the main asset entry. If the main asset entry doesn’t correspond to a real file, then the asset with the lowest resolution is used as the fallback for devices with device pixel ratios below that resolution. The entry should still be included in the pubspec.yaml manifest, however.

**Comment by andrewkolos** [comment]
The fix for has been cherry-picked onto the stable channel. Expect it to became available with the next 3.10 hotfix release.
